capirca/lib/versa.py

```diff
@@ -0,0 +1,877 @@
+          if not valid_addrs:
+            logging.warning(
+                'WARNING: Term %s has 0 valid source IPs, skipping.', term.name)
+            continue
+          term.source_address = valid_addrs
+          for addr in term.source_address:
+            addr_counter=self._BuildAddressBook('src-'+term.name,
+                                               addr_counter, addr)
+
+        # Filter destination_address based on filter_type & add to address book
+        if term.destination_address:
+          valid_addrs = []
+          for addr in term.destination_address:
+            if addr.version in self._AF_MAP[filter_type]:
+              valid_addrs.append(addr)
+          if not valid_addrs:
+            logging.warning(
+                'WARNING: Term %s has 0 valid destination IPs, skipping.',
+                term.name)
+            continue
+          term.destination_address = valid_addrs
+          for addr in term.destination_address:
+            addr_counter=self._BuildAddressBook('dest-'+term.name,
+                                                addr_counter, addr)
+
+        new_term = Term(term, self.from_zone, self.to_zone,
+                                         self.addressbook, verbose)
+        new_terms.append(new_term)
+
+        # we need to remove icmp from the protocol and add it to the
+        #pan-application list
+        if term.protocol and 'icmp' in ' '.join(term.protocol):
+          term.protocol.remove('icmp')
+          term.versa_application.append('ICMP')
+        # Because Versa terms can contain inet and inet6 addresses. We have to
+        # have ability to recover proper AF for ICMP type we need.
+        # If protocol is empty or we cannot map to inet or inet6 we insert bogus
+        # af_type name which will cause new_term.NormalizeIcmpTypes to fail.
+
+        #if not term.protocol:
+        #  icmp_af_type = 'unknown_af_icmp'
+        #else:
+        #  icmp_af_type = self._AF_ICMP_MAP.get(
+        #      term.protocol[0], 'unknown_af_icmp')
+        #tmp_icmptype = new_term.NormalizeIcmpTypes(
+        #    term.icmp_type, term.protocol, icmp_af_type)
+
+        # NormalizeIcmpTypes returns [''] for empty, convert to [] for eval
+        #normalized_icmptype = tmp_icmptype if tmp_icmptype != [''] else []
+        # rewrites the protocol icmpv6 to icmp6
+        if 'icmpv6' in term.protocol:
+          protocol = list(term.protocol)
+          protocol[protocol.index('icmpv6')] = 'icmp6'
+        else:
+          protocol = term.protocol
+        new_application_set = {'sport': self.BuildPort(term.source_port),
+                               'dport': self.BuildPort(term.destination_port),
+                               'protocol': protocol }
+
+        # add this only of one of the parameters is not None
+        if ( new_application_set['sport'] or new_application_set['dport'] or
+            new_application_set['protocol'] ):
+          for application_set in self.applications:
```

Comment:
Confirming that Versa treats applications separately, and you need both applications and versa applications (like paloalto)